python/tests/io/test_pyarrow.py

```diff
@@ -1330,3 +1342,411 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    assert datafile.upper_bounds[6] == STRUCT_INT64.pack(5)
+    assert datafile.upper_bounds[7] == STRUCT_INT64.pack(6)
+
+
+def test_metrics_mode_non_default_trunc() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    table_metadata.properties["write.metadata.metrics.default"] = "truncate(2)"
+    fill_parquet_file_metadata(
+        datafile,
+        metadata,
+        len(file_bytes),
+        table_metadata,
+    )
+
+    assert len(datafile.value_counts) == 5
+    assert len(datafile.null_value_counts) == 5
+    assert len(datafile.nan_value_counts) == 0
+
+    assert len(datafile.lower_bounds) == 5
+    assert datafile.lower_bounds[1].decode() == "aa"
+    assert datafile.lower_bounds[2] == STRUCT_FLOAT.pack(1.69)[:2]
+    assert datafile.lower_bounds[5] == STRUCT_INT64.pack(1)[:2]
+    assert datafile.lower_bounds[6] == STRUCT_INT64.pack(1)[:2]
+    assert datafile.lower_bounds[7] == STRUCT_INT64.pack(2)[:2]
+
+    assert len(datafile.upper_bounds) == 5
+    assert datafile.upper_bounds[1].decode() == "zz"
+    assert datafile.upper_bounds[2] == STRUCT_FLOAT.pack(100)[:2]
+    assert datafile.upper_bounds[5] == STRUCT_INT64.pack(9)[:2]
+    assert datafile.upper_bounds[6] == STRUCT_INT64.pack(5)[:2]
+    assert datafile.upper_bounds[7] == STRUCT_INT64.pack(6)[:2]
+
+
+def test_column_metrics_mode() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    table_metadata.properties["write.metadata.metrics.default"] = "truncate(2)"
+    table_metadata.properties["write.metadata.metrics.column.strings"] = "none"
+    table_metadata.properties["write.metadata.metrics.column.list.element"] = "counts"
+    fill_parquet_file_metadata(
+        datafile,
+        metadata,
+        len(file_bytes),
+        table_metadata,
+    )
+
+    assert len(datafile.value_counts) == 4
+    assert len(datafile.null_value_counts) == 4
+    assert len(datafile.nan_value_counts) == 0
+
+    assert len(datafile.lower_bounds) == 3
+    assert datafile.lower_bounds[2] == STRUCT_FLOAT.pack(1.69)[:2]
+    assert 5 not in datafile.lower_bounds
+    assert datafile.lower_bounds[6] == STRUCT_INT64.pack(1)[:2]
+    assert datafile.lower_bounds[7] == STRUCT_INT64.pack(2)[:2]
+
+    assert len(datafile.upper_bounds) == 3
+    assert datafile.upper_bounds[2] == STRUCT_FLOAT.pack(100)[:2]
+    assert 5 not in datafile.upper_bounds
+    assert datafile.upper_bounds[6] == STRUCT_INT64.pack(5)[:2]
+    assert datafile.upper_bounds[7] == STRUCT_INT64.pack(6)[:2]
+
+
+def test_offsets() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    fill_parquet_file_metadata(datafile, metadata, len(file_bytes), table_metadata)
+
+    assert datafile.split_offsets is not None
+    assert len(datafile.split_offsets) == 1
+    assert datafile.split_offsets[0] == 4
+
```

Comment:
Thanks for all the tests, this is great @maxdebayser!

Could you add these two for my sanity:

```python
def test_write_and_read_stats_schema(table_schema_nested: Schema):
    tbl = pa.Table.from_pydict({
        "foo": ["a", "b"],
        "bar": [1, 2],
        "baz": [False, True],
        "qux": [["a", "b"], ["c", "d"]],
        "quux": [[("a", (("aa", 1), ("ab", 2)))], [("b", (("ba", 3), ("bb", 4)))]],
        "location": [[(52.377956, 4.897070), (4.897070, -122.431297)],
                     [(43.618881, -116.215019), (41.881832, -87.623177)]],
        "person": [("Fokko", 33), ("Max", 42)]  # Possible data quality issue
    },
        schema=schema_to_pyarrow(table_schema_nested)
    )
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    visited_paths = []

    def file_visitor(written_file: Any) -> None:
        visited_paths.append(written_file)

    with tempfile.TemporaryDirectory() as tmpdir:
        pq.write_to_dataset(tbl, tmpdir, file_visitor=file_visitor)

    assert visited_paths[0].metadata.num_columns == len(stats_columns)

def test_stats_types(table_schema_nested: Schema):
    stats_columns = pre_order_visit(table_schema_nested, PyArrowStatisticsCollector(table_schema_nested, {}))

    # the field-ids should be sorted
    assert all(stats_columns[i].field_id <= stats_columns[i + 1].field_id for i in range(len(stats_columns)-1))
    assert [col.iceberg_type for col in stats_columns] == [
        StringType(),
        IntegerType(),
        BooleanType(),
        StringType(),
        StringType(),
        StringType(),
        IntegerType(),
        FloatType(),
        FloatType(),
        StringType(),
        IntegerType(),
    ]

```